make_prg/recursion_tree.py

```diff
@@ -0,0 +1,404 @@
+
+            if not has_a_single_interval:
+                interval_subalignments = NodeFactory._partition_alignment_into_interval_subalignments(alignment,
+                                                                                                      all_intervals)
+                return MultiIntervalNode(nesting_level, alignment, parent_node, prg_builder, interval_subalignments)
+
+        # if parent is multi cluster, children should not be
+        skip_building_multi_cluster_node = isinstance(parent_node, MultiClusterNode)
+        if not skip_building_multi_cluster_node:
+            clustering_result = kmeans_cluster_seqs(alignment, min_match_length)
+            cluster_further = NodeFactory._infer_if_we_should_cluster_further(alignment, clustering_result)
+
+            if cluster_further:
+                cluster_subalignments = NodeFactory._get_subalignments_by_clustering(alignment, clustering_result)
+                return MultiClusterNode(nesting_level, alignment, parent_node, prg_builder,
+                                        cluster_subalignments)
+
+        # here, the node is a leaf
+        return LeafNode(nesting_level, alignment, parent_node, prg_builder)
+
+    #####################################################################################################
+    #  helper methods
+    @staticmethod
+    def _alignment_has_issues(alignment: MSA) -> bool:
+        num_unique_nongapped_seqs = get_number_of_unique_ungapped_sequences(alignment)
+        too_few_unique_sequences = num_unique_nongapped_seqs <= 2
+        if too_few_unique_sequences:
+            return True
+
+        num_unique_gapped_seqs = get_number_of_unique_gapped_sequences(alignment)
+
+        # this is an assert as if it happens, something very wrong with the two previous functions is happening
+        assert num_unique_nongapped_seqs <= num_unique_gapped_seqs
+
+        alignment_has_ambiguity = num_unique_nongapped_seqs < num_unique_gapped_seqs
+        if alignment_has_ambiguity:
+            # TODO: fix alignment by deduplicating nongapped seqs and realigning them
+            # TODO: the annoying thing is that this deduplicated nongapped alignment will differ from the input alignment
+            # TODO: but I think it can be an ok solution
+            return True
+
+        return False
+
+    @staticmethod
+    def _get_nesting_level(parent_node: Optional[RecursiveTreeNode]) -> int:
+        is_root = parent_node is None
+        if is_root:
+            nesting_level = 0
+        else:
+            need_to_go_down_one_level = isinstance(parent_node, MultiClusterNode)
+            if need_to_go_down_one_level:
+                nesting_level = parent_node.nesting_level + 1
+            else:
+                nesting_level = parent_node.nesting_level
+        return nesting_level
+    #####################################################################################################
```

Comment:
I propose removing this, see comment above

Reply:
done